### ISSUE TITLE
Autoencoder output inconsistent with TensorFlow

### ISSUE THREAD

**Issue by ddaspit** [plaidml/plaidml#83]
I would like to use PlaidML, because it support AMD GPUs. Unfortunately, I have had issues trying to replicate the results for an autoencoder model that I implemented using the TensorFlow backend in Keras. The following is a simplified version of an autoencoder model that I implemented, which still demonstrates the issue:
``` python
import cv2
import numpy

import plaidml.keras
plaidml.keras.install_backend()

from keras.models import Model
from keras.layers import Input
from keras.layers.convolutional import Conv2D, MaxPooling2D, UpSampling2D
from keras.optimizers import Adam
from keras.datasets import cifar10

if __name__ == "__main__":
    input_ = Input(shape=(32, 32, 3))

    x = input_
    x = Conv2D(64, (3, 3), activation='relu', padding='same')(x)
    x = MaxPooling2D((2, 2), padding='same')(x)
    x = Conv2D(128, (3, 3), activation='relu', padding='same')(x)
    x = MaxPooling2D((2, 2), padding='same')(x)
    x = Conv2D(256, (3, 3), activation='relu', padding='same')(x)
    x = MaxPooling2D((2, 2), padding='same')(x)

    x = Conv2D(256, (3, 3), activation='relu', padding='same')(x)
    x = UpSampling2D((2, 2))(x)
    x = Conv2D(128, (3, 3), activation='relu', padding='same')(x)
    x = UpSampling2D((2, 2))(x)
    x = Conv2D(64, (3, 3), activation='relu', padding='same')(x)
    x = UpSampling2D((2, 2))(x)
    x = Conv2D(3, (3, 3), padding='same', activation='sigmoid')(x)

    autoencoder = Model(input_, x)
    optimizer = Adam(lr=5e-5, beta_1=0.5, beta_2=0.999)
    autoencoder.compile(optimizer=optimizer, loss='mean_absolute_error')

    (x_train, y_train), (x_test, y_test) = cifar10.load_data()
    x_train = x_train.astype('float32') 
    x_test = x_test.astype('float32')
    x_train /= numpy.max(x_train)
    x_test /= numpy.max(x_test)

    for epoch in range(50):
        index = epoch * 32
        batch = x_train[index:index + 32]
        loss = autoencoder.train_on_batch(batch, batch)
        print(loss)

    test_batch = x_test[0:14]
    figure = numpy.stack([
        test_batch,
        autoencoder.predict(test_batch),
    ], axis=1)

    figure = figure.reshape((2, 7) + figure.shape[1:])
    images_shape = numpy.array(figure.shape)
    n = len(images_shape)
    if n % 2 == 0:
        y_axes = list(range(1, n - 1, 2))
        x_axes = list(range(0, n - 1, 2))
    else:
        y_axes = list(range(0, n - 1, 2))
        x_axes = list(range(1, n - 1, 2))
    new_axes = y_axes, x_axes, [n - 1]
    new_shape = [numpy.prod(images_shape[x]) for x in new_axes]
    figure = numpy.transpose(
        figure,
        axes=numpy.concatenate(new_axes)
        ).reshape(new_shape)

    figure = numpy.clip(figure * 255, 0, 255).astype('uint8')

    cv2.imwrite('output.jpg', figure)
```
When I run this code using the PlaidML backend in Keras, I get the following output:
![plaidml_output](https://user-images.githubusercontent.com/3261883/35518810-295c438a-0545-11e8-936c-4edbc679ef32.jpg)
The images on the left are the input and the images on the right are the output. As you can see, the output images are almost completely black.

When I run the same code using the TensorFlow backend in Keras, I get the following output:
![tf_output](https://user-images.githubusercontent.com/3261883/35518889-5f69b4f8-0545-11e8-9ec9-e027368e654a.jpg)
This output is more what I would expect from the autoencoder.

These results were produced on a Windows 10 machine with an AMD RX 580 GPU.

Any help would be appreciated to help me understand why I am getting different results between PlaidML and TensorFlow. I truly appreciate the work that has been done to develop PlaidML.

**Comment by tzerrell** [comment]
Thanks for the report! Can you try running the training in TensorFlow, saving the weights, loading the weights in PlaidML, and running inference to see if the result is correct? This should help us narrow down the problem to training or inference.

Unfortunately we have a number of high priority issues on our plate right now and we will not be able to work on a fix immediately. But thank you again for letting us know about this problem so we can put fixing it in queue.

**Comment by ddaspit** [comment]
As suggested, I tried training with TensorFlow and predicting with PlaidML. The results were correct. To verify, I tried the opposite, training with PlaidML and predicting with TensorFlow. As expected, the results were incorrect. So it looks like, it is a problem with training.

**Comment by tzerrell** [comment]
Thanks for running those tests! It's good to know that the problem is in training. We'll look into a fix.

**Comment by unlugi** [comment]
Hello, have you found a solution to your problem? I am training a conv-autoencoder as well, but my reconstructions are black images similar to the ones you have shown above. Thank you in advance.

**Comment by tzerrell** [comment]
We have reproduced the error in somewhat simpler test case (a necessary condition for the error to occur is using mean absolute error as the loss function), but we do not yet have a fix. I am actively working on fixing this so I'll claim this issue to make that clear.

**Comment by tzerrell** [comment]
Just wanted to give a quick update that we're working on a release that includes a fix for this, and we will update this issue when that release is out.

**Comment by tzerrell** [comment]
Our 0.3.5 release includes a release for this. Let us know if the problem persists for you.
